### ISSUE TITLE
`verify_strict` and weak-key forgery

### ISSUE THREAD

**Issue by rozbb** [dalek-cryptography/ed25519-dalek#259]
This issue comes after a long time thinking about what `verify_strict` does and whether we want to remove it. I argue (surprise!) that we should not.

The `verify_strict` function checks that the verifying key `A`, and `sig.R` are not low-order points. 

The [original intent](https://github.com/dalek-cryptography/ed25519-dalek/pull/98) seems to have been a stronger check, namely checking torsion-freeness of those curve points, but that never happened afaict. These are very different checks.

Regardless, let's assume it is doing torsion-freeness checks. There is documentation in the [README](https://github.com/dalek-cryptography/ed25519-dalek/tree/f036eaf48297b09efa251b9c41052c8132b9c0cb#a-note-on-signature-malleability) and in [`verify_strict`](https://docs.rs/ed25519-dalek/1.0.1/ed25519_dalek/struct.PublicKey.html#method.verify_strict) claiming that permitting `R` or `A` with torsion (which still have large order) yields signature malleability. I don't think this is true though, as both the `R` and `A` components are included in the hash `H(R || A || M)`, and `A` is still large enough order that the change in the hash yields an entirely different challenge "in the exponent". The docs are discussed more [here](https://github.com/dalek-cryptography/ed25519-dalek/issues/130). There's a similar claim in [this comment](https://github.com/dalek-cryptography/ed25519-dalek/issues/20#issuecomment-564274006), but again I haven't seen an example yet.

However! I think `verify_strict` is still valuable.

The low-order check on `A` means **`verify_strict` provides resistance to weak-key forgeries**. In particular, if `A` is low-order, then `H(R || A || M) A == H(R || A || M') A` with 1/8 probability, for any distinct messages `M` and `M'`. We have a demonstration of this in our [tests](https://github.com/dalek-cryptography/ed25519-dalek/pull/153/files). In terms of impact, Scuttlebutt had a weak-key forgery vuln due to precisely this issue (see [here](https://eprint.iacr.org/2019/526.pdf), section 7.1). The attack context is: the attacker can pick their pubkey and provide a signature but the message is not known to them. The attacker uses a weak pubkey `A`, generates an `s`, lets `R = sB`, and sends `(A, σ=(s, R))` to the verifier. With 1/8 probability, the secret message `M` known to the verifier satisfies `sB == R + H(R || A || M) A`. Note that the forgery method here can also be thought of as a malleability attack, since you can find plenty of `(s, R=sB)` pairs that pass verification on `(A, M)` by simple trial and error.

Some things that aren't clear to me are whether there really is a malleability issue on large-order pubkeys with torsion, and whether the low-order check on `R` does anything for us. @isislovecruft if you could explain that'd be awesome.

Going forward, we can do a few things:

1. Keep `verify_strict` and update the docs around it, explaining weak-key forgery and removing previous claims about malleability. According to the paper, Go's NaCl impl, Cloudflare's CIRCL, and Libsodium have all implemented low-order pubkey checks.
2. Delete it and put a warning about weak-key forgery. Recommend to anyone concerned about this that they should use sigs over prime order curves. Like [RedDSA](https://github.com/ZcashFoundation/reddsa), [schnorrkel](https://github.com/w3f/schnorrkel), or [jq255e](https://docs.rs/crrl/latest/crrl/jq255e/index.html).
3. Make `verify_strict` actually do the torsion-freeness checks. There's actually a reason you might want this. If you did this in batch verif as well, then it would agree perfectly with single verif. This way you could have consistent batch verif that rejects weak-key forgeries.

I like (1), especially since you can't really do a low-order check in our current API. IMO `verify_strict` should be the default functionality, but I don't think we can do that now.

(3) sounds OK too, though allegedly the torsion-freeness check is pretty expensive and it might not make sense in a batch setting from the start.

cc @tarcieri

**Comment by tarcieri** [comment]
@rozbb if `verify_strict` provides behavior that matches* libsodium and other libraries, I agree it's worth keeping around

*I know this has been exhaustively studied but I'm not sure what might've changed in the interim since those blog posts and papers were written

**Comment by rozbb** [comment]
Libsodium [does](https://github.com/jedisct1/libsodium/blob/24211d370a9335373f0715664271dfe203c7c2cd/src/libsodium/crypto_sign/ed25519/ref10/open.c#L44) small order checks on `A` and `R`. It also does canonicity checks which we also do but only accidentally.

But I don't think [HACL(https://github.com/hacl-star/hacl-star/blob/main/code/ed25519/Hacl.Impl.Ed25519.Verify.fst) does low-order checks. Go's [ed25519 impl](https://cs.opensource.google/go/go/+/master:src/crypto/ed25519/ed25519.go;l=290) does not (and they also have a faulty scalar check). Go's NaCl [just uses](https://cs.opensource.google/go/x/crypto/+/master:nacl/sign/sign.go;drc=04dced17f71c03d747bac91c3b2623de124f1e75;l=66) their ed25519 impl. And [CIRCL](https://github.com/cloudflare/circl/blob/459b64f53bc65c6ef18923652e59d8d28da75057/sign/ed25519/ed25519.go#L322) doesn't do the checks.

So I guess no one actually fixed this. I'm still pro keeping `verify_strict`. What do you think?

**Comment by rozbb** [comment]
Ok cool. There's a related thing I'll make a PR for: the `legacy_compatibility` feature disable scalar checks even for `verify_strict`. This is bad imo. We should just disable `verify_strict` if that feature is set.

**Comment by rozbb** [comment]
Closing. We're keeping `verify_strict` and updating docs.

**Comment by rozbb** [comment]
Documenting here:

I think `verify_strict` implicitly tests canonical encoding for `R` and `A`. The docs mention this in passing [here](https://github.com/dalek-cryptography/ed25519-dalek/blob/928d6d15f8cdce9e2fa6d3ab9f63361e663f245b/src/verifying.rs#L55).

My reasoning, previously stated [here](https://github.com/dalek-cryptography/ed25519-dalek/issues/267#issuecomment-1398006456), is that it you need to know `dlog(R)` in order to make a signature, and the only non-canonical points with known discrete log are the low-order points. This goes for `A` as well. Thus, checking for low-order is sufficient to check non-canonicity, assuming hardness of discrete log on those specific points.

**Comment by FiloSottile** [comment]
> Go's [ed25519 impl](https://cs.opensource.google/go/go/+/master:src/crypto/ed25519/ed25519.go;l=290) does not (and they also have a faulty scalar check).

Uuuuh I'd like to hear more about the scalar check? :)

**Comment by rozbb** [comment]
Hi Fillipo! I just looked into it and I think I was mistaken. What I saw was [this line](https://cs.opensource.google/go/go/+/master:src/crypto/ed25519/ed25519.go;l=309)
```go
if len(sig) != SignatureSize || sig[63] & 224 != 0 {
    return false
}
```
This is the old libsodium scalar check, which only checks that the top 3 bits are unset, and leads to possible signature malleability.

But I see now that you later call `edwards25519.NewScalar.SetCanonicalBytes(sig[32:])`, which properly checks that `isReduced` is `true`.

So nevermind, sorry about that!
